File: OpenTkCSharp/WinFormsOpenGL31/Chapter007/015_ColoredCube/Camera.cs
using System;
using OpenTK;
using OpenTK.Graphics.OpenGL;

class Camera
{
    private Vector3 _eyePos;
    private Vector3 _targetPos;
    private Matrix4 _projMatrix;
    private Matrix4 _viewMatrix;
    private int _uProjMatrixLoc;
    private int _uViewMatrixLoc;
    private Vector3 _orientation = new Vector3(0f, 1f, 0f);
    private float _nearPlane = 0.1f;
    private float _farPlane = 1000f;

    public Camera(int program, Vector3 eyePos, Vector3 targetPos)
    {
        _eyePos = eyePos;
        _targetPos = targetPos;

        _uProjMatrixLoc = GL.GetUniformLocation(program, "uProjMatrix");
        _uViewMatrixLoc = GL.GetUniformLocation(program, "uViewMatrix");
        if (_uProjMatrixLoc < 0 || _uViewMatrixLoc < 0)
        {
            Console.WriteLine("Failed to get the storage location of uProjMatrix or uViewMatrix");
            return;
        }

        SetViewProjection();
    }

    public void SetViewProjection()
    {
        _projMatrix = Matrix4.CreatePerspectiveFieldOfView(MathHelper.DegreesToRadians(30f), 1f, _nearPlane, _farPlane);
        _viewMatrix = Matrix4.LookAt(_eyePos, _targetPos, _orientation);

        GL.UniformMatrix4(_uProjMatrixLoc, false, ref _projMatrix);
        GL.UniformMatrix4(_uViewMatrixLoc, false, ref _viewMatrix);
    }
}

File: glMatrixTypeScript/ch07/015_colored-cube/src/Camera.ts
import { mat4, vec3 } from "gl-matrix";

export class Camera
{
    private _gl: WebGLRenderingContext;
    private _eyePos: vec3;
    private _targetPos: vec3;
    private _projMatrix: mat4;
    private _viewMatrix: mat4;
    private _uProjMatrixLoc: WebGLUniformLocation;
    private _uViewMatrixLoc: WebGLUniformLocation;
    private _orientation = vec3.fromValues(0, 1, 0);
    private _nearPlane = 0.1;
    private _farPlane = 100;

    public constructor(gl: WebGLRenderingContext, program: WebGLProgram, eyePos: vec3, targetPos: vec3)
    {
        this._gl = gl;
        this._eyePos = eyePos;
        this._targetPos = targetPos;

        this._projMatrix = mat4.create();
        this._viewMatrix = mat4.create();

        this._uProjMatrixLoc = this._gl.getUniformLocation(program, "uProjMatrix");
        this._uViewMatrixLoc = this._gl.getUniformLocation(program, "uViewMatrix");
        if (!this._uProjMatrixLoc || !this._uViewMatrixLoc)
        {
            console.log("Failed to get the storage location of uProjMatrix or uViewMatrix");
            return;
        }

        this.SetViewProjection();
    }

    public SetViewProjection(): void
    {
        mat4.perspective(this._projMatrix, 30.0 * (Math.PI / 180.0), 1, this._nearPlane, this._farPlane);
        mat4.lookAt(this._viewMatrix, this._eyePos, this._targetPos, this._orientation);

        this._gl.uniformMatrix4fv(this._uProjMatrixLoc, false, this._projMatrix);
        this._gl.uniformMatrix4fv(this._uViewMatrixLoc, false, this._viewMatrix);
    }
}